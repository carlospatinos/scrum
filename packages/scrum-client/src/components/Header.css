.indigo {
    background-color: #3f51b5 !important;
}

.navbar {
    font-weight: 300;
    box-shadow: 0 2px 5px 0 rgb(0 0 0 / 16%), 0 2px 10px 0 rgb(0 0 0 / 12%);
    font-family: "Roboto", sans-serif;
}