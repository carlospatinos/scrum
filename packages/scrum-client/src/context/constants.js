const ACTIONS_FETCH = 'FETCH';
const ACTIONS_POST = 'POST';
const ACTIONS_FAILURE = 'FAILURE';
const ACTIONS_SUCCESS = 'SUCCESS';
const LIBRARY_ROOT = 'SCRUM';
const ROUTE_CHANGED = 'ROUTE_CHANGED';

export {
  ACTIONS_FETCH,
  ACTIONS_POST,
  ACTIONS_FAILURE,
  ACTIONS_SUCCESS,
  LIBRARY_ROOT,
  ROUTE_CHANGED,
};
