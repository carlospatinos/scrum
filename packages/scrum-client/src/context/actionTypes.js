import ContextUtil from './utils';
import { ACTIONS_POST } from './constants';

const LOGIN_ACTIONS = ContextUtil.generateActionTypes({
  featureName: 'LOGIN',
  restVerb: ACTIONS_POST,
});
const LOGOUT_ACTIONS = ContextUtil.generateActionTypes({
  featureName: 'LOGOUT',
});

export { LOGIN_ACTIONS, LOGOUT_ACTIONS };
