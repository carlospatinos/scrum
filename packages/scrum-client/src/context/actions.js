/* eslint-disable no-console */
import { END_POINTS } from 'scrum-common';
import { API_CONSTANTS } from '../constants';

export async function loginUser(dispatch, loginPayload) {
  const requestOptions = {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    // 'Access-Control-Allow-Credentials': true,
    body: JSON.stringify(loginPayload),
  };

  try {
    dispatch({ type: 'REQUEST_LOGIN' });
    const response = await fetch(
      `${API_CONSTANTS.API_BASE_URL}${END_POINTS.AUTH}${END_POINTS.AUTH_LOCAL}`,
      requestOptions
    );
    const data = await response.json();

    if (data.user) {
      dispatch({ type: 'LOGIN_SUCCESS', payload: data });
      localStorage.setItem(API_CONSTANTS.CURRENT_USER, JSON.stringify(data));
      return data;
    }

    dispatch({ type: 'LOGIN_ERROR', error: data.message });
  } catch (error) {
    dispatch({ type: 'LOGIN_ERROR', error });
    console.log(error);
  }
  return undefined;
}

export async function login3ppUser(dispatch) {
  const requestOptions = {
    method: 'GET',
    headers: { 'Content-Type': 'application/json' },
    // 'Access-Control-Allow-Credentials': true,
    credentials: 'include',
  };

  try {
    dispatch({ type: 'REQUEST_LOGIN' });
    const response = await fetch(
      `${API_CONSTANTS.API_BASE_URL}${END_POINTS.AUTH}${END_POINTS.LOGIN_SUCCESS}`,
      requestOptions
    );

    const data = await response.json();
    if (data.user) {
      dispatch({ type: 'LOGIN_SUCCESS', payload: data });
      localStorage.setItem(API_CONSTANTS.CURRENT_USER, JSON.stringify(data));
      return data;
    }

    dispatch({ type: 'LOGIN_ERROR', error: data.message });
    console.log(data.message);
  } catch (error) {
    dispatch({ type: 'LOGIN_ERROR', error });
    console.log(error);
  }
  return undefined;
}

export async function logout(dispatch) {
  dispatch({ type: 'LOGOUT' });
  localStorage.removeItem(API_CONSTANTS.CURRENT_USER);
  localStorage.removeItem(API_CONSTANTS.ACCESS_TOKEN_NAME);
  // TODO is this needed? localStorage.clear();
  localStorage.clear();
}
