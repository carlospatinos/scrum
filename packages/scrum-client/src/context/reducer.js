/* eslint-disable no-shadow */
import { CURRENT_USER } from '../constants/apiConstants';

const user = localStorage.getItem(CURRENT_USER)
  ? JSON.parse(localStorage.getItem(CURRENT_USER)).user
  : '';
const token = localStorage.getItem(CURRENT_USER)
  ? JSON.parse(localStorage.getItem(CURRENT_USER)).login_access_token
  : '';

export const initialState = {
  user: '' || user,
  login_access_token: '' || token,
  loading: false,
  errorMessage: null,
};

export const AuthReducer = (initialState, action) => {
  switch (action.type) {
    case 'REQUEST_LOGIN':
      return {
        ...initialState,
        loading: true,
      };
    case 'LOGIN_SUCCESS':
      return {
        ...initialState,
        user: action.payload.user,
        login_access_token: action.payload.login_access_token,
        loading: false,
      };
    case 'LOGOUT':
      return {
        ...initialState,
        user: '',
        login_access_token: '',
      };

    case 'LOGIN_ERROR':
      return {
        ...initialState,
        loading: false,
        errorMessage: action.error,
      };

    default:
      throw new Error(`Unhandled action type: ${action.type}`);
  }
};
