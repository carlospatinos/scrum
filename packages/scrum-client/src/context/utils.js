import { ACTIONS_FETCH, ACTIONS_SUCCESS, ACTIONS_FAILURE, LIBRARY_ROOT } from './constants';
/**
 * Generates an action type by concatenating the library, feature name and key.
 *
 * @param {String} library
 * @param {String} featureName
 * @param {String} key
 */
const generateActionType = (library, featureName, key) => `${library}/${featureName}/${key}`;

/**
 * Generates a partial implementation of `generateActionType` fixing library and featureName.
 *
 * @param {String} library
 * @param {String} featureName
 */
const generateFeatureActionType = (library, featureName) => key =>
  generateActionType(library, featureName, key);

/**
 * Generates the three default action types, baseAction,  successAction and failureAction of a given feature.
 *
 * @param {*} { library = LIBRARY_ROOT, featureName, restVerb = ACTIONS_FETCH }
 * @returns
 */
const generateActionTypes = ({ library = LIBRARY_ROOT, featureName, restVerb = ACTIONS_FETCH }) => {
  const baseActionKey = `${restVerb}_${featureName.toUpperCase()}`;
  const featureActionType = generateFeatureActionType(library, featureName);
  const baseAction = featureActionType(baseActionKey);
  const successAction = featureActionType(`${baseActionKey}_${ACTIONS_SUCCESS}`);
  const failureAction = featureActionType(`${baseActionKey}_${ACTIONS_FAILURE}`);
  return { baseAction, successAction, failureAction };
};
/**
 * Generates a context action, it handles the three action types of a given feature action.
 * - dispatchFxn( type: action.baseAction): Before invoking the `apiFxn`/
 * - dispatchFxn( type: action.successAction): When the response of  `apiFxn` is received.
 * - dispatchFxn( type: action.baseAction):  When an error  of `apiFxn` is received.
 *
 * @param {Function} dispatchFxn - The dispatcher function.
 * @param {Function} apiFxn - The API function to invoke.
 * @param {Object} action - The baseAction, successAction and failureAction of a given feature.
 */
const generateAction = (dispatchFxn, apiFxn, action) => async req => {
  const initialAction = {
    type: action.baseAction,
    payload: {},
  };
  dispatchFxn(initialAction);

  try {
    const response = await apiFxn(req);
    console.log('response.data', response.data);
    dispatchFxn({
      type: action.successAction,
      payload: response.data,
    });
    return response;
  } catch (e) {
    dispatchFxn({
      type: action.failureAction,
      error: e.message,
    });
    return undefined;
  }
};

/**
 * Context utility , it generates actions types and actions.
 */
const ContextUtil = { generateActionType, generateActionTypes, generateAction };
export default ContextUtil;
