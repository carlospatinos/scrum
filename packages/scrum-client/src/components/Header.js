import React from 'react';
import { Navbar, Nav, NavDropdown, Button, Figure } from 'react-bootstrap';
import { useTranslation } from 'react-i18next';
import { useLocation } from 'react-router-dom';
import { PATHS } from '../constants';
import routes from '../config/routes';
import { useAuthState } from '../context';
import './Header.css';

const visibleRoute = isUserLoggedIn => {
  const privacy = isUserLoggedIn ? route => route.isPrivate : route => !route.isPrivate;
  return route => privacy(route) && route.isVisibleOnMenu;
};

export default function Header() {
  const userDetails = useAuthState();
  const location = useLocation();
  const isLoggedIn = userDetails.login_access_token;
  const visibleRouteCriteria = visibleRoute(isLoggedIn);
  const { t } = useTranslation();

  // TODO to a function
  const image =
    userDetails.user.profileImageUrl === undefined
      ? '/icons/default-profile.png'
      : userDetails.user.profileImageUrl;

  return (
    <header>
      <div>
        <Navbar variant="dark" className="bg-primary fixed-top navbar-expand-md" expand="lg">
          <Navbar.Brand className="" href={PATHS.HOME}>
            <img src="/appLogo.svg" alt={t(`Header.altLogo`)} />
            <strong className="align-middle">{t(`Header.titleBrand`)}</strong>
          </Navbar.Brand>
          <Navbar.Toggle aria-controls="basic-navbar-nav" />
          <Navbar.Collapse id="basic-navbar-nav">
            <Nav className="ml-auto" activeKey={location.pathname}>
              {routes.filter(visibleRouteCriteria).map(route =>
                route.path === PATHS.PROFILE ? (
                  <NavDropdown
                    key={route.path}
                    title={t(`Routes.${route.title}`)}
                    id="basic-nav-dropdown"
                  >
                    <Figure variant="top" className="d-flex justify-content-center">
                      <Figure.Image
                        roundedCircle
                        width={70}
                        height={70}
                        alt={t(`Header.altProfilePic`)}
                        src={image}
                      />
                    </Figure>
                    <div className="d-flex justify-content-center">
                      {userDetails.user.firstName} {userDetails.user.lastName}
                    </div>
                    <div className="d-flex justify-content-center">
                      <small className="text-secondary">{userDetails.user.email}</small>
                    </div>
                    <NavDropdown.Item eventKey="manage" className="d-flex justify-content-center">
                      <Button
                        variant="outline-secondary"
                        size="sm"
                        className="rounded-pill"
                        disabled
                      >
                        <strong>{t('Header.btnManageYourAccount')}</strong>
                      </Button>
                    </NavDropdown.Item>
                    <NavDropdown.Divider />
                    <Nav.Link
                      key={PATHS.LOGOUT}
                      href={PATHS.LOGOUT}
                      className="d-flex justify-content-center"
                    >
                      <Button variant="outline-secondary" size="sm" className="rounded">
                        {t('Routes.logout')}
                      </Button>
                    </Nav.Link>
                    <NavDropdown.Divider />
                    <NavDropdown.Item eventKey="policy">
                      <small className="text-secondary">{t('Header.lblPrivacyPolicy')}</small>
                    </NavDropdown.Item>
                    <NavDropdown.Item eventKey="terms">
                      <small className="text-secondary">{t('Header.lblTermsAndConditions')}</small>
                    </NavDropdown.Item>
                  </NavDropdown>
                ) : (
                  <Nav.Link key={route.path} href={route.path} active={route.isActive}>
                    <strong>{t(`Routes.${route.title}`)}</strong>
                  </Nav.Link>
                )
              )}
            </Nav>
          </Navbar.Collapse>
        </Navbar>
      </div>
    </header>
  );
}
