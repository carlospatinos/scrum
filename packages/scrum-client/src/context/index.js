import { loginUser, login3ppUser, logout } from './actions';
import { AuthProvider, useAuthDispatch, useAuthState } from './context';

export { AuthProvider, useAuthState, useAuthDispatch, loginUser, login3ppUser, logout };
