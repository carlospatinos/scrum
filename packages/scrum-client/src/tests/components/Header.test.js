import { render, cleanup } from '../test-utils';
import Header from '../../components/Header';
import { useAuthState } from '../../context/context';
import { basicUser } from '../mocks/user';

jest.mock('../../context/context');
afterEach(cleanup);

describe('Header', () => {
  test('To match snapshot', () => {
    //   mock hook call
    useAuthState.mockReturnValue(basicUser);

    // render the component
    const { asFragment } = render(<Header />);

    // validates vs snapshot
    expect(asFragment()).toMatchSnapshot();
  });
});
