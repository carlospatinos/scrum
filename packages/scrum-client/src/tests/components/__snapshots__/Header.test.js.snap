// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header To match snapshot 1`] = `
<DocumentFragment>
  <header>
    <div>
      <nav
        class="bg-primary fixed-top navbar-expand-md navbar navbar-expand-lg navbar-dark"
      >
        <a
          class="navbar-brand"
          href="/home"
        >
          <img
            alt="Logo"
            src="/appLogo.svg"
          />
          <strong
            class="align-middle"
          >
            Scrum
          </strong>
        </a>
        <button
          aria-controls="basic-navbar-nav"
          aria-label="Toggle navigation"
          class="navbar-toggler collapsed"
          type="button"
        >
          <span
            class="navbar-toggler-icon"
          />
        </button>
        <div
          class="navbar-collapse collapse"
          id="basic-navbar-nav"
        >
          <div
            class="ml-auto navbar-nav"
          >
            <a
              class="nav-link"
              data-rb-event-key="/home"
              href="/home"
            >
              <strong>
                Home
              </strong>
            </a>
            <div
              class="dropdown nav-item"
            >
              <a
                aria-expanded="false"
                aria-haspopup="true"
                class="dropdown-toggle nav-link"
                href="#"
                id="basic-nav-dropdown"
                role="button"
              >
                Profile
              </a>
            </div>
            <a
              class="nav-link"
              data-rb-event-key="/planningconfig"
              href="/planningconfig"
            >
              <strong>
                Planning Config
              </strong>
            </a>
          </div>
        </div>
      </nav>
    </div>
  </header>
</DocumentFragment>
`;
